### DISCUSSION TITLE
2023-08-11 - Testing 2 - How do you test your work? - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by lukecookssw** [SSWConsulting/SSW.Rules.Content#6285] [Chewing the Fat]
Hey testers!

Let's talk about [this rule](https://www.ssw.com.au/rules/acceptance-criteria/)

**Comment by jernejk** [comment]
In my previous job ~8-9 years ago, we had dedicated testers and we had really good processes for testing.

At SSW client projects, there were a few projects where we had testing team, otherwise we do more extensive testing before we push to UAT/production to make sure the project quality.

Regarding acceptance criteria, I think ChatGPT might be best approach for consistent format. Developers might not consider/know how detailed they might need to make acceptance criteria so that manual testers could make tests.

**Comment by brydeno** [comment]
I find the most important thing is to have a goal for any testing that occurs. That helps focus the testing. For instance Apples would appear to be "Make a feature delightful". That makes it clear to everyone the focus is on UX and that drives the don't add extra features.

**Reply by wicksipedia** [reply]
Just like a unit test - ~do~ test one thing and ~do~ test it well

**Comment by drwharris** [comment]
I find that either developers are reluctant to speak up about the true effort required to test properly or the business is unwilling to spend the appropriate amount of time on building good code coverage.

**Comment by leahy268** [comment]
I don't have a lot of experience with automated testing.  Almost everything I do requires manual testing by hand.  It's a pain as I would really love a way to automatically test bicep and YAML but I usually only find problems after I commit.

**Comment by jakebayliss** [comment]
Maybe there should be a standard of 1 to * between an acceptance criteria item and tests, given the AC is testable. This would likely ensure both the acceptance criteria are written in a way that we can easily confirm a pass (with tests) and also easily lay out the tests we need

**Comment by danielmackay** [comment]
When writing user stories I like to distinguish between **Acceptence Criteria** and **Dev Notes*.  Acceptence Criteria are functional changes to the software.  Dev Notes, are points the may need to keep in mind during development, or more technical concerns that do not have a functional impact.  These are often noted during refinement sessions.

If the Acceptance Criteria are really complex, I also like to use the Gerkin Syntax (Given/When/Then).  This makes the Acceptance Criteria very explicit, but the down side is that they take a lot longer to document.

**Comment by tkapa** [comment]
Standing up and implementing testing structures is almost always the reason why writing test cases gets left behind. There are  also times when testing goes overboard (e.g. 1 million + Unit tests). 

In concept I like this rule. However, I don't really love the given example of Acceptance Criteria, it reads more like PBI Requirements or a User Story. The rule also feels like it could be broken into smaller, more digestible rules.

**Comment by ozdemir-mehmet** [comment]
My project building is all the testing I need! (just kidding). I really like writing unit tests again FakeXRMEasy for Dynamics testing, makes life so much easier knowing I have the same (faked) data and schema I'm testing my code against.

**Comment by bahjatalaadelssw** [comment]
The TV team does exhaustive manual testing on every product we produce and spends a considerable amount of time doing random exploratory testing developed by the product owner.

**Comment by AlexRothwell** [comment]
I find Acceptance Criteria quite important when the description of a PBI contains excessive amounts of detail. The ACs help to narrow down what the focus of the PBI actually is and what might be partially describing the existing state. Of course there's always a range of quality of ACs and developers should understand how and when to give feedback on these to improve outcomes

**Comment by chrisschultzssw** [comment]
I haven't had to do a lot with testing, but it's useful to learn about for when I'm writing scripts, etc.

**Comment by Ben0189** [comment]
The titles of the rules confuse me where I think something like "Better way of writing Acceptance Criteria" will be more appropriate to the context in the rules

**Comment by sethdaily** [comment]
I'm learning more about testing than I'd ever known. I can understand the importance of having good testing to have efficient feedback loops, save time, and prevent buggy projects from going live.

**Comment by jomelsontanco** [comment]
In my previous waterfall projects, we not only have a dedicate test team, we have multiple dedicated test teams in **each** testing phase. We have also detailed step-by-step test scripts documented for each scenario. My move to an Agile project was a culture shock testing-wise as there were no longer any dedicated test teams. The test scripts are replaced by a simplistic acceptance criteria. All the developers almost do all of the testing. We used to joke that this is not Agile but Fragile.

At SSW FireBootCamp, I include most of my tests in my PR referencing the acceptance criteria as welll as any tasks in the originating PBI. In the PBI, I then refer back to the PR to let the PO know that the tests reside in the PR. The Scrum Team would also test our work during Sprint review asking us to demonstrate in real-time some requested scenarios.

**Comment by stephanfakossw** [comment]
Nice summary...wonder what the next topic is 😊

**Comment by Dhruv-0987** [comment]
Testing and code coverage is extremely important and a crucial part of deployment which most dev teams Ignore. Devs are okay until the main functionality is working and ignore the edge cases which should not be the case. Testing should be an official requirement in the definition of Done especially in our internal projects.

**Reply by wicksipedia** [reply]
I fixed a typo in this comment, should've done more testing 😂

**Comment by cvionnet** [comment]
Great rule

**Comment by jeoffreyfischer** [comment]
I haven't done any testing yet, but this is a good reminder of the importance of acceptance criteria

**Comment by SamBWagner** [comment]
Automated testing is awesome and is a great way of ensuring that your code works the way you think it should. That said, code coverage is a terrible metric for defining if tests are good/valuable.

**Comment by AntPolkanov** [comment]
I got a bit confused. The rule is about Acceptance Criteria, while CTF was half about AC and half about testing. Technically, devs should write tests for ACs, but these two seem to be really broad topics to mix together.

**Comment by apforsyth** [comment]
This was valuable from a non-developer point of view because it makes us focus on defining Acceptance criteria more carefully - not just a set of tasks but and outcome with specific functionality

**Comment by Harry-Ross** [comment]
I thought the rule was important in clarifying the importance of acceptance criteria for projects. Also keen to learn more about how we can use automated testing to make projects better! I've spent some time with Jest in personal projects, and think that it can be a great tool for ensuring that code is performing its desired tasks.

**Comment by Jord-Gui** [comment]
Good to know more on the Acceptance Criteria and Definition of Done.

**Comment by zacharykeeping** [comment]
Really well written rule - a good summary of Acceptance Criteria and ensuring PBIs have well defined, testable end point.

**Comment by tombui99** [comment]
In my opinion, a detailed Acceptance Criteria is absolutely necessary that can tremendously help the devs to build/fix and test a feature/bug. With that in mind, it is debatable whether all PBIs require automated unit testing as it means more devs time spent.

**Comment by KahaMason** [comment]
Tests are super valuable to have. I think baking them into the definition of done is a good way of ensuring writing tests for new code is not forgotten. If you're primarily focusing on developing features and pushing them, it can be easy to overlook writing tests which can build to become a massive amount of technical debt.

Also with some projects with like 1million+ tests, it can be good to see what is actually being tested and if the tests actually provide value instead of writing tests for the sake of writing tests.

**Comment by MattParkerDev** [comment]
Good clarification around writing good acceptance criteria

**Comment by JeanThirion** [comment]
Love the Rule, well done ⭐

**Comment by amankumarrr** [comment]
Nice explanation on writing good acceptance criteria and its distinction with the definition of done.

**Comment by steven0x51** [comment]
Good acceptance criterias are important for clarifying the purpose of the PBI

**Comment by PennyWalker** [comment]
I think it's really important to get this right, so I'm glad we're covering it, but I didn't get to learn a huge amount about testing this time around - I missed having a video to learn from! :D

**Comment by AttackOnMorty** [comment]
Acceptance criteria should be **testable**. If not, it may indicate that the story is:
- Not clear enough -> Elaborate more
- Too complex -> Break down into testable units
- Too technical -> Involve devs

**Comment by JackDevAU** [comment]
✅Testing is important
✅Acceptance criteria even more so 
✅The rule does a good job of driving home the importance of discussions between the PO and the dev team. 


Having a detailed PBI with AC makes development streamlined. Without having this its always a guessing game and more often than not, wasted dev time going back and forth.

**Comment by jackreimers** [comment]
Going back and checking past PBIs as part of this CTF was a good reminder of the need to be testing code, it's easy to just get the PBI done and move on to the next.

**Comment by 2fernandez** [comment]
**Testing** is a very critical stage in UX Design. Being a UX Designer, I can only speak to this.

**User Testing** - validates some of the stories and use cases we formulate at the research level. 
Unfortunately, most projects do not have detailed enough foundational research phases (spec reviews) or a User Testing phase, often because the client is not committed to proper procedure, mainly due to perception or budget. 

❌ The solution itself, in some instances, suffers.
❌ Most of the time, the client and the solution providers operate in the land of innuendo and personal bias. 
❌ No accurate data or user testing is looked at; we are invisible to user performance.
❌ There is no feedback loop on improvement

**Comment by leehawkinsssw** [comment]
Good acceptance criteria help both developers and testers. Before work starts on a PBI, the ACs need to be reviewed, especially to ensure that they are testable! Checking the ACs in the delivered PBI is then the starting point for the testers and, if they're met, they can continue on to performing deeper testing.

**Comment by ricksu978** [comment]
Suggestions 
To be more specific on "how" to setup the test data.

1. > When I enter ‘Adam’ in the Search box and click ‘Search’ I will see all entries starting with Adam in the Grid

   Given there are 10 entries in the Grid and 3 of them starting with 'Adam'
  When I enter 'Adam' in the Search box and click 'Search'
  Then I will see only 3 entries starting with 'Adam' displayed in the Grid.

2. > When I enter ‘zzz’ in the Search box and click ‘Search’ I will see no entries in the Grid

   Given there are 10 entries in the Grid and none of them starting with 'zzz'
  When I enter 'zzz' in the Search box and click 'Search'
  Then I will see 0 entries displayed in the Grid.

**Comment by piers-sinclair** [comment]
I only skimmed it but the general ideas look good, however I don't think you should ever have a PBI with no Acceptance Criteria. It encourages bad practices.

In my opinion, I think the structure and story of the rule could be improved.

At the moment it feels like there is a main story at the top and then it became a bunch of tips and ancillary information below. These tips and ancillary information should be built into the story or split into their own rules. I quickly lost interest after reading the main intro

Additionally, there is a violation of DRY at the bottom of the rule because it overlaps with https://www.ssw.com.au/rules/document-discoveries

**Comment by Marxoz** [comment]
Good Acceptance Criteria is so important for testing. If test cases can be written BEFORE development starts, developers can write code and test it already at the same time. That will require the acceptance criteria to be written before the PBI is pulled into a sprint, and the creation of the steps being included in the estimate.

**Comment by babakamyljanovssw** [comment]
I don't have experience on testing yet. But having Acceptance Criteria in PBI helped me to understand clearly what is required from dev to build/fix feature and saved a lot of time.

**Comment by sylhuang** [comment]
Good Acceptance Criteria could save devs and PO tons of time clarifying what should be achieved in a PBI. It also helps with TDD.

**Comment by CoffeeBugHub** [comment]
I acquired knowledge about user stories and their corresponding acceptance criteria. Surprisingly, in actual projects, I have yet to encounter one with well-defined and substantial acceptance criteria for a PBI.

**Comment by KristenHu** [comment]
Good AC should be readable, especially for non-dev, it'll help us to add automated tests and manual tests.

**Comment by jimmidier** [comment]
I agree with @ricksu978 that if ACs are the source of truth for both development and testing they should have consistent format such as the BDD styled one @ricksu978  mentioned. It's easier to write for POs and more readable for devs and testers in that way.

**Comment by tino-liu** [comment]
As long as the users can read it and be "Oh! Easy peasy" and at the same time, devs can read it and go "Oh! I know how to write unit and manual tests for this!", that would be the definition of a good AC.

**Comment by Aibono1225** [comment]
I like the concept of acceptance criteria, it's a good guideline that make sure everyone knows what's needed and prevent misunderstandings.

**Comment by zl24622** [comment]
As a developer, I find good accepetance criteria makes the development process eaiser, no need to  think too much about UX, functionality, more focuing on technical implementation.

For automated testing and unit test code coverage, not sure how it's being implemented on low-code development.

**Comment by johnxu16** [comment]
Having good acceptance criteria makes sure testers do not miss essential bits and keep developers on the rail.

Even if extracting relatively common and volatile modules into automated testing is time-consuming, it's worth for keeping the quality of the project when it tends to be a gigantic monster in the future.

**Comment by jerryluossw** [comment]
In most of my client projects, there will be black box testing by professional testers and acceptance testing by product managers or project managers will be required at the end.

In addition, I totally agree that accepetance criteria is so important, good accepetance criteria will unify the ideas of all involved team members, so that when the final acceptance will not cause the phenomenon of squabbling.

**Comment by christoment** [comment]
Acceptance criteria definitely helps to guide in which part we need to tests and how to write the code itself 😊

In topic of writing tests to fulfil acceptance criteria: I find writing tests in legacy projects is painful, but necessary... Not unusual I was faced with scenario where writing small tests require the same amount of effort it takes to write the code in such projects...

**Comment by BrookJeynes** [comment]
Solid acceptance criteria helps further down the line such as helping with testing and allows a clear guideline of when a feature is considered done.

**Comment by vladislav-kir** [comment]
Acceptance criteria helps keep everyone on the same page regarding what needs to be coded and tested for the PBI. It is definitely important.

**Comment by tomek-i** [comment]
As port of the .NET 8 migration our Team decided to add actual tests in because the current project's test suite is ... well, not worth mentioning at all.

**Comment by Abf09** [comment]
Seems like a useful rule specially where it's clarified the difference between Definition of Done and Acceptance criteria, I have not yet tested my current project, but I do look forward to running automated tests once the work on Back end is completed for fireBootCamp.

**Comment by jontygardner** [comment]
Tailoring this rule for the world of video production!

Acceptance Criteria are like a script for user stories in video production. They define what needs to be done for a story to be complete, just like a script outlines scenes in a film. Developers follow them to build the story's functionality, and testers use them to create relevant tests. Think of them as clear shooting instructions for a video. Just as directors collaborate with their team, Product Owners work with developers to avoid unnecessary "nice-to-haves" and prioritize essential tasks. If technical details arise, the team discusses them together. Acceptance Criteria act as the "Rejection Tests," ensuring the story's goals are met, similar to editing rounds for videos.

**Comment by JackDuan123** [comment]
Accecptance Criteria is very useful!
Acceptance Criteria is easier for the developers to work the tickets.
However, in the real case, we need to figure out what's crucial, what's not when writting acceptance Criteria, is a small issue worth a few days of fix?

**Comment by 2fernandez** [comment]
I learnt something new today.
The difference between "Acceptance Criteria" and "Definition of Done".

Now for the meme of the week: 
![meme-test-code-in-production-live-dangerously](https://github.com/SSWConsulting/SSW.Rules.Content/assets/27670973/c2dee745-27f1-44b6-b907-92671d3a1d9a)

**Comment by matt-goldman** [comment]
This rule is great. I'm a big fan of Gerkin Syntax as using Specflow it lets you create tests that test directly against your acceptance criteria. Most of the tests we write as developers are tests to prevent someone else from breaking our code, rather than testing that the product meets the specified functional requirements.

**Comment by Hona** [comment]
All these things can harmonise very well with processes but I find usually you get a weak point or two that can remove some value.

e.g.

- Not every team has strong tech leadership to refine small items for devs
- Not every team has a PO/SM ensuring there are small PBIs & that the business value is there

**Comment by ozair-ashfaque** [comment]
Acceptance Criteria helps to determine how we are going to write the code and which area of code requires testing
